File: src/pkg04_service_cpp/src/server_node.cpp
// server_node.cpp
// Exposes the service /add_two_ints. When a client sends two integers,
// the server replies with their sum.

#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "pkg03_tutorial_interfaces/srv/add_two_ints.hpp"

using AddTwoInts = pkg03_tutorial_interfaces::srv::AddTwoInts;

class ServerNode : public rclcpp::Node
{
public:
  ServerNode() : Node("add_two_ints_server")
  {
    service_ = this->create_service<AddTwoInts>(
      "add_two_ints",
      std::bind(&ServerNode::handle_request, this,
                std::placeholders::_1, std::placeholders::_2));

    RCLCPP_INFO(this->get_logger(), "Service /add_two_ints is ready.");
  }

private:
  void handle_request(
    const std::shared_ptr<AddTwoInts::Request> request,
    std::shared_ptr<AddTwoInts::Response> response)
  {
    response->sum = request->a + request->b;
    RCLCPP_INFO(this->get_logger(),
                "Request: a=%ld, b=%ld -> sum=%ld",
                request->a, request->b, response->sum);
  }

  rclcpp::Service<AddTwoInts>::SharedPtr service_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ServerNode>());
  rclcpp::shutdown();
  return 0;
}

File: src/pkg05_service_py/pkg05_service_py/server_node.py
#!/usr/bin/env python3
"""Service server in Python. Adds two integers."""

import rclpy
from rclpy.node import Node
from pkg03_tutorial_interfaces.srv import AddTwoInts


class ServerNode(Node):
    def __init__(self):
        super().__init__('add_two_ints_server_py')

        # create_service(srv_type, srv_name, callback)
        self.srv = self.create_service(
            AddTwoInts, 'add_two_ints', self.handle_request)

        self.get_logger().info('Service /add_two_ints is ready (Python)')

    def handle_request(self, request, response):
        # The framework gives us a pre-allocated response object.
        # Fill its fields and return it.
        response.sum = request.a + request.b
        self.get_logger().info(
            f'Request: a={request.a}, b={request.b} -> sum={response.sum}')
        return response


def main(args=None):
    rclpy.init(args=args)
    node = ServerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
